File: prover.pl

add(N, 0, N).
add(N, suc(M), suc(Y)) :- add(N, M, Y).

dec(0, 0).
dec(suc(N), N).

sub(N, 0, N).
sub(N, suc(M), U) :- sub(N, M, Y), dec(Y, U).

base([], []).
base([pair(_, NNF)|T], [NNF|CT]) :- base(T, CT).

dash(A, 0, A).
dash(A, suc(N), [N|A]).

dump([], []).
dump([H|T], A) :- dump(T, S), dash(S, H, A).

free(pre(_, _, V), V).
free(con(P, Q), V) :- free(P, F1), free(Q, F2), append(F1, F2, V).
free(dis(P, Q), V) :- free(P, F1), free(Q, F2), append(F1, F2, V).
free(uni(P), V) :- free(P, F), dump(F, V).
free(exi(P), V) :- free(P, F), dump(F, V).

over(S, _, 0, S).
over(_, H, suc(_), H).

more(X, S, H, 0, Y) :- sub(X, H, XS), over(S, H, XS, Y).
more(_, _, H, suc(_), Y) :- dec(H, Y).

mend(_, _, [], []).
mend(X, S, [H|T], [MO|ME]) :- sub(H, X, XS), more(X, S, H, XS, MO), mend(X, S, T, ME).

subst(X, S, pre(B, I, V), pre(B, I, W)) :-  mend(X, S, V, W).
subst(X, S, con(P, Q), con(PP, QQ)) :- subst(X, S, P, PP), subst(X, S, Q, QQ).
subst(X, S, dis(P, Q), dis(PP, QQ)) :-  subst(X, S, P, PP), subst(X, S, Q, QQ).
subst(X, S, uni(P), uni(PP)) :-  subst(suc(X), suc(S), P, PP).
subst(X, S, exi(P), exi(PP)) :-  subst(suc(X), suc(S), P, PP).

fresh([], 0).
fresh([H|T], suc(Y)) :- fresh(T, A), dec(A, B), sub(B, H, C), add(C, H, Y).

frees([], []).
frees([H|T], Y) :-  free(H, F), frees(T, FF), append(F, FF, Y).

stop(C, _, [], C).
stop(C, P, [H|T], Y) :- P = H -> Y = [] ; stop(C, P, T, Y).

negate(0, 1).
negate(1, 0).

track(S, _, pre(B, I, V), Y) :-
  append(S, [pair(0, pre(B, I, V))], SS),
  base(S, BASE),
  negate(B, NB),
  stop([SS], pre(NB, I, V), BASE, Y).
track(S, _, con(P, Q), [SP, SQ]) :- 
  append(S, [pair(0, P)], SP),
  append(S, [pair(0, Q)], SQ).
track(S, _, dis(P, Q), [SS]) :-
  append(S, [pair(0, P), pair(0, Q)], SS).
track(S, _, uni(P), [Y]) :- 
  base(S, B),
  frees([uni(P)|B], F),
  fresh(F, FF),
  subst(0, FF, P, SFF),
  append(S, [pair(0, SFF)], Y).
track(S, N, exi(P), [Y]) :- 
  subst(0, N, P, SP),
  append(S, [pair(0, SP), pair(suc(N), exi(P))], Y).

solve([], [[]]).
solve([pair(F, S)|T], Y) :- track(T, F, S, Y).

solves([], []).
solves([H|T], Y) :- solve(H, HS), solves(T, TS), append(HS, TS, Y).

prover([], 1).
prover([H|T], Y) :- solves([H|T], S), prover(S, Y).

check(P, Y) :- prover([[pair(0, P)]], Y).

:- begin_tests(cases).
test(add) :- add(suc(suc(0)), suc(suc(suc(0))), suc(suc(suc(suc(suc(0)))))).
test(base) :- base([pair(1,2)], [2]).
test(dash0) :- dash([suc(0)], 0, [suc(0)]).
test(dash1) :- dash([suc(0)], suc(0), [0, suc(0)]).
test(dump) :- dump([0, suc(suc(0)), suc(0)], [suc(0), 0]).
test(free) :- 
  A = con(pre(1, 0, [suc(0)]), pre(1, 0, [suc(0)])),
  free(A,[suc(0), suc(0)]).
test(fresh) :- fresh([suc(0), suc(suc(0)), 0], suc(suc(suc(0)))).
% TODO: test
% over, mend, more, subst, frees, stop, track, solve, solves, prover
test(track_dis) :-
  T = dis(pre(0, 0, [0]), pre(1, 0, [0])), 
  track([], 0, T, Y),
  Y = [[
    pair(0, pre(0, 0, [0])), 
    pair(0, pre(1, 0, [0]))
  ]].
test(check_simple) :-
  T = dis(
    pre(0, 0, [0]),
    pre(1, 0, [0])
  ),
  check(T, 1).
test(check) :-
  T =
    dis(
      uni(con(
        pre(0, 0, [0]),
        pre(0, suc(0), [0])
      )),
      dis(
        exi(pre(1, suc(0), [0])),
        exi(pre(1, 0, [0]))
      )
    ),
  check(T, 1).
:- end_tests(cases).

% :-trace(track).
% :-trace.
% :-debug.
:- run_tests.

File: prover.hs

import Prelude ((==), (/=), (<), (<=), (>=), (>), (+), (-), (*), (/), (**),
  (>>=), (>>), (=<<), (&&), (||), (^), (^^), (.), ($), ($!), (++), (!!), Eq, Show,
  error, id, return, not, fst, snd, map, filter, concat, concatMap, reverse,
  zip, null, takeWhile, dropWhile, all, any, Integer, negate, abs, divMod,
  String, Bool(True, False), Maybe(Nothing, Just));
import qualified Prelude;

data Nat = ZeroNat | Suc Nat deriving Prelude.Show;

equalNat :: Nat -> Nat -> Bool;
equalNat ZeroNat (Suc x2) = False;
equalNat (Suc x2) ZeroNat = False;
equalNat (Suc x2) (Suc y2) = equalNat x2 y2;
equalNat ZeroNat ZeroNat = True;

instance Eq Nat where {
  a == b = equalNat a b;
};

data Nnf = Pre Bool Nat [Nat] | Con Nnf Nnf | Dis Nnf Nnf | Uni Nnf | Exi Nnf deriving Prelude.Show;


add :: Nat -> Nat -> Nat;
add x ZeroNat = x;
add x (Suc n) = Suc (add x n);

dec :: Nat -> Nat;
dec ZeroNat = ZeroNat;
dec (Suc n) = n;

sub :: Nat -> Nat -> Nat;
sub x ZeroNat = x;
sub x (Suc n) = dec (sub x n);

base :: [(Nat, Nnf)] -> [Nnf];
base [] = [];
base (h : t) = snd h : base t;

dash :: [Nat] -> Nat -> [Nat];
dash l ZeroNat = l;
dash l (Suc n) = n : l;

dump :: [Nat] -> [Nat];
dump [] = [];
dump (h : t) = dash (dump t) h;

free :: Nnf -> [Nat];
free (Pre uu uv v) = v;
free (Con p q) = free p ++ free q;
free (Dis p q) = free p ++ free q;
free (Uni p) = dump (free p);
free (Exi p) = dump (free p);

over :: Nat -> Nat -> Nat -> Nat;
over s uu ZeroNat = s;
over uv h (Suc uw) = h;

more :: Nat -> Nat -> Nat -> Nat -> Nat;
more x s h ZeroNat = over s h (sub x h);
more uu uv h (Suc uw) = dec h;

mend :: Nat -> Nat -> [Nat] -> [Nat];
mend uu uv [] = [];
mend x s (h : t) = more x s h (sub h x) : mend x s t;

subst :: Nat -> Nat -> Nnf -> Nnf;
subst x s (Pre b i v) = Pre b i (mend x s v);
subst x s (Con p q) = Con (subst x s p) (subst x s q);
subst x s (Dis p q) = Dis (subst x s p) (subst x s q);
subst x s (Uni p) = Uni (subst (Suc x) (Suc s) p);
subst x s (Exi p) = Exi (subst (Suc x) (Suc s) p);

fresh :: [Nat] -> Nat;
fresh [] = ZeroNat;
fresh (h : t) = Suc (add (sub (dec (fresh t)) h) h);

frees :: [Nnf] -> [Nat];
frees [] = [];
frees (h : t) = free h ++ frees t;

equalNnf :: Nnf -> Nnf -> Bool;
equalNnf (Uni x4) (Exi x5) = False;
equalNnf (Exi x5) (Uni x4) = False;
equalNnf (Dis x31 x32) (Exi x5) = False;
equalNnf (Exi x5) (Dis x31 x32) = False;
equalNnf (Dis x31 x32) (Uni x4) = False;
equalNnf (Uni x4) (Dis x31 x32) = False;
equalNnf (Con x21 x22) (Exi x5) = False;
equalNnf (Exi x5) (Con x21 x22) = False;
equalNnf (Con x21 x22) (Uni x4) = False;
equalNnf (Uni x4) (Con x21 x22) = False;
equalNnf (Con x21 x22) (Dis x31 x32) = False;
equalNnf (Dis x31 x32) (Con x21 x22) = False;
equalNnf (Pre x11 x12 x13) (Exi x5) = False;
equalNnf (Exi x5) (Pre x11 x12 x13) = False;
equalNnf (Pre x11 x12 x13) (Uni x4) = False;
equalNnf (Uni x4) (Pre x11 x12 x13) = False;
equalNnf (Pre x11 x12 x13) (Dis x31 x32) = False;
equalNnf (Dis x31 x32) (Pre x11 x12 x13) = False;
equalNnf (Pre x11 x12 x13) (Con x21 x22) = False;
equalNnf (Con x21 x22) (Pre x11 x12 x13) = False;
equalNnf (Exi x5) (Exi y5) = equalNnf x5 y5;
equalNnf (Uni x4) (Uni y4) = equalNnf x4 y4;
equalNnf (Dis x31 x32) (Dis y31 y32) = equalNnf x31 y31 && equalNnf x32 y32;
equalNnf (Con x21 x22) (Con y21 y22) = equalNnf x21 y21 && equalNnf x22 y22;
equalNnf (Pre x11 x12 x13) (Pre y11 y12 y13) =
  x11 == y11 && equalNat x12 y12 && x13 == y13;

stop :: [[(Nat, Nnf)]] -> Nnf -> [Nnf] -> [[(Nat, Nnf)]];
stop c uu [] = c;
stop c p (h : t) = if equalNnf p h then [] else stop c p t;

track :: [(Nat, Nnf)] -> Nat -> Nnf -> [[(Nat, Nnf)]];
track s uu (Pre b i v) =
  stop [s ++ [(ZeroNat, Pre b i v)]] (Pre (not b) i v) (base s);
track s uv (Con p q) = [s ++ [(ZeroNat, p)], s ++ [(ZeroNat, q)]];
track s uw (Dis p q) = [s ++ [(ZeroNat, p), (ZeroNat, q)]];
track s ux (Uni p) =
  [s ++ [(ZeroNat, subst ZeroNat (fresh (frees (Uni p : base s))) p)]];
track s n (Exi p) = [s ++ [(ZeroNat, subst ZeroNat n p), (Suc n, Exi p)]];

solve :: [(Nat, Nnf)] -> [[(Nat, Nnf)]];
solve [] = [[]];
solve (h : t) = track t (fst h) (snd h);

solves :: [[(Nat, Nnf)]] -> [[(Nat, Nnf)]];
solves [] = [];
solves (h : t) = solve h ++ solves t;

nulla :: [a] -> Bool;
nulla [] = True;
nulla (uu : uv) = False;

main' ::
  (([[(Nat, Nnf)]] -> Bool) ->
    ([[(Nat, Nnf)]] -> [[(Nat, Nnf)]]) -> [[(Nat, Nnf)]] -> Bool) ->
    Nnf -> Bool;
main' a p = a nulla solves [[(ZeroNat, p)]];

test :: Nnf;
test =
  Dis (Uni (Con (Pre False ZeroNat [ZeroNat])
              (Pre False (Suc ZeroNat) [ZeroNat])))
    (Dis (Exi (Pre True (Suc ZeroNat) [ZeroNat]))
      (Exi (Pre True ZeroNat [ZeroNat])));

iterator :: (a -> Bool) -> (a -> a) -> a -> Bool;
iterator g f c = if g c then True else iterator g f (f c);

check :: Nnf -> Bool;
check = main' iterator;

-- let p = Dis (Pre True ZeroNat [ZeroNat]) (Pre False ZeroNat [ZeroNat])
-- track ([(ZeroNat, p)]) ZeroNat p


-- let pp = [(ZeroNat,Pre False ZeroNat [ZeroNat])], Pre True ZeroNat [ZeroNat] => []
